### PR TITLE
Element: export new React 18 APIs

### PR DESCRIPTION
**Pull request by jsnajdr**
Starts exporting new React 18 API from `@wordpress/element`. A standalone PR we'll later use in #46467 (`createRoot` and concurrent mode) and in #46606 (`useSelect` with `useSyncExternalStore`).

### CODE REVIEW

File: packages/element/src/react-platform.js [hunk 1]
```diff
@@ -43,6 +44,20 @@ export { render };
  */
```
Comment (tyxla) [inline packages/element/src/react-platform.js:44]: Should we add a `@deprecated` for `render()` and `hydrate()` now that they are deprecated in React 18?

File: packages/element/README.md [hunk 2]
```diff
@@ -198,6 +198,14 @@ _Returns_
 
 -   `Object`: Ref object.
 
+### createRoot
+
+Creates a new React root for the target DOM node.
```
Comment (tyxla) [inline packages/element/README.md:203]: An ideal place to mention that these are deprecated now in React 18.

File: packages/element/src/react.js [hunk 3]
```diff
@@ -185,6 +186,11 @@ export { useRef };
  */
 export { useState };
 
+/**
+ * @see https://reactjs.org/docs/hooks-reference.html#usesyncexternalstore
+ */
+export { useSyncExternalStore };
```
Comment (tyxla) [inline packages/element/src/react.js:212]: Does it make sense to add the rest of the new hooks (`useId`, `useTransition`, `useDeferredValue`, `useInsertionEffect`)? I can immediately see opportunities for almost all of them. 

Note that if we add `useTransition` we might want to expose `startTransition` as well.

File: packages/element/src/react-platform.js [hunk 4]
```diff
@@ -43,6 +44,20 @@ export { render };
...
+ */
+export { createRoot };
+
+/**
+ * Creates a new React root for the target DOM node and hydrates it with a pre-generated markup.
+ *
+ * @see https://reactjs.org/docs/react-dom-client.html#hydrateroot
+ */
+export { hydrateRoot };
+
```
Comment (tyxla) [inline packages/element/src/react-platform.js:60]: Should we also expose `Suspense`?

**Review by tyxla — COMMENTED** [review]
Thanks for this one! I wonder if you'd like to add a few more of the new APIs that we'll likely use or at least experiment with in the near future.

File: packages/element/src/react-platform.js (hunk 1 above)
Comment (youknowriad) [reply packages/element/src/react-platform.js:44]: I think we if add @deprecated, we need to also add a real deprecated function call, so let's leave it for later when we migrate all of our usage.

File: packages/element/src/react.js (hunk 3 above)
Comment (jsnajdr) [reply packages/element/src/react.js:212]: Good idea, pushed a commit that exports them all.

File: packages/element/src/react-platform.js (hunk 4 above)
Comment (jsnajdr) [reply packages/element/src/react-platform.js:60]: `Suspense` has already been exposed for some time.

File: packages/element/src/react-platform.js (hunk 1 above)
Comment (tyxla) [reply packages/element/src/react-platform.js:44]: Sure. Maybe adding a comment would suffice then.

File: packages/element/src/react-platform.js (hunk 4 above)
Comment (tyxla) [reply packages/element/src/react-platform.js:60]: Oh, right 👍

**Review by tyxla — APPROVED** [review]
This looks great! Thanks 🚀

**Pull request merged**